### ISSUE TITLE
API: Consider separate contracts for Maps/Records

### ISSUE THREAD

**Issue by DLehenbauer** [microsoft/tiny-calc#76]
Some _potential_ differences between maps/records that may motivate a separate contract:
* The set of keys in a map is dynamic (i.e., no fixed schema)
* It is not an error to read non-existent keys (returns undefined)
* You can test for the presence of a key, get a count of entries, and enumerate map entries, key, values.
* You can delete keys and clear a map.

I think we could probably expose these as an extension of the record interfaces:

```ts
interface IMapReader<TKey, TValue> extends IReader<Record<TKey, TValue | undefined>> {
    has(key: TKey): boolean;
    entries(): IEnumarable<[TKey, TValue]>;
    keys(): IEnumarable<TKey>;
    values(): IEnumarable<TValue>;
    readonly size: number;
}

interface IMapWriter<TKey, TValue> extends IWriter<Record<TKey, TValue>> {
    delete(key: TKey): void;
    clear(): void;
}
```

**Comment by DLehenbauer** [comment]
...or is a Record just a Map with an immutable shape?

```ts
interface IShapeReader {
    has(key: TKey): boolean;
    keys(): IEnumarable<TKey>;
    readonly size: number;
}
```

```ts
interface IShapeWriter {
    delete(key: TKey): void;
    clear(): void;
}
```

**Comment by DLehenbauer** [comment]
Addressed in #78
